docs/proposals/02-backupinfra.md

```diff
@@ -0,0 +1,232 @@
+
+#### Deletion
+- On shoot deletion, the BackupEntry instance i.e. shoot specific instance will get deletion timestamp because of ownerReference.
+- If `deletionGracePeriod` configured in GCM component configuration is expired, BackupInfrastructure Controller will delete the backup folder associated with it from backup object store.
+- Finally, it will remove the `finalizer` from backupEntry instance.
+
+### Alternative
+
+![sequence-diagram](Backup-Infrastructure-Provisioning-with-deletion-job.svg)
+
+## Discussion points / variations
+### Manual vs dynamic bucket creation
+- As per limit observed on different cloud providers, we can have single bucket for backups on one cloud providers. So, we could avoid the little complexity introduced in above approach by pre-provisioning buckets as a part of landscape setup. But there won't be anybody to detect bucket existence and its reconciliation.
+
+- Another thing we can have is, we can let administrator register the pool of root backup infra resource and let the controller schedule backup on one of this.
+
+- One more variation here could be to create bucket dynamically per hash of shoot UID.
+
+
+### SDK vs Terraform
+Initial reason for going for terraform script is its stability and the provided parallelism/concurrency in resource creation. For backup infrastructure, Terraform scripts are very minimal right now. Its simply have bucket creation script. With shared bucket logic, if possible we might want to isolate access at [directory](#terminology) level but again its additional one call. So, we will prefer switching to SDK for all object store operations.
+
+### Limiting the number of shoots per bucket
+We can have central map of configuration in `gardener-controller-component-configuration.yaml`.
+Where we will mark supported count of shoots per cloud provider. Most probable space could be,
+`controller.backupInfrastructures.quota`.
+
+## Backward compatibility
+### Migration
+- Create shoot specific folder.
+- Transfer old objects.
+- Create manifest of objects on new bucket
+    - Each entry will have status: None,Copied, NotFound.
+    - Copy objects one by one.
+- Scale down etcd-main with old config. :warning: Cluster down time
+- Copy remaining objects
+- Scale up etcd-main with new config.
+- Destroy Old bucket and old backup namespace. It can be immediate or preferrably __lazy__ deletion.
+
+![backup-migration-sequence-diagram](./Backup-Migration.svg)
+
+### Legacy Mode alternative
+- If Backup namespace present in seed cluster, then follow the legacy approach.
+- i.e. reconcile creation/existence of shoot specific bucket and backup namespace.
+- If backup namespace is not created, use shared bucket.
+- __Limitation__ Never know when the existing cluster will be deleted, and hence, it might be little difficult to maintain with next release of gardener. This might look simple and straight-forward for now but may become pain point in future design change if required.
```

Comment:
Which future design change are you referring to?

How would the migration approach look like/be done (via script, automated in Gardener, ...)?

Reply:
> Which future design change are you referring to?

Nothing specific. Mostly not. But we never know in worst case when some new use case pops up and we have to refactor the design.

> How would the migration approach look like/be done (via script, automated in Gardener, ...)?

This will be automated part in gardener. New step in shoot-reconcile flow just before `DeployEtcd`. No script. That will be removed in some next release.

Reply:
Well, what shall I say.. challenging ;-) So the alternative approach won't be taken (legacy mode) but everything will be migrated. OK, thanks.